File: credcrypt/credcrypt.go
// Package credcrypt protects credentials that would otherwise travel in
// cleartext over a plain HTTP connection.
//
// # Threat model
//
// A passive network attacker who can read request bodies — a shared Wi-Fi
// segment, a mirroring switch port, a logging reverse proxy a few hops away —
// can otherwise lift a username and password straight out of a login POST.
//
// # Scheme
//
// The client asks for a handshake, generates a random AES-256-GCM key,
// encrypts the request payload with it, and wraps that key with RSA-OAEP
// (SHA-256) under the server's ephemeral public key. Only the holder of the
// matching private key — the server itself — can recover the payload, so the
// password is never observable on the wire.
//
// Ciphertext alone would however be replayable verbatim by that same passive
// attacker, so a handshake also yields a single-use challenge, and the client
// puts it inside the encrypted payload. The server accepts a challenge exactly
// once, only for the scope it was issued for, and only until it expires. A
// captured envelope is therefore worthless afterwards.
//
// # Scope
//
// A "scope" is the operation a challenge was issued for (login, signup, users,
// share, share-unlock). It stops an envelope captured from one endpoint from
// being replayed against another, and is checked after decryption so that a
// mismatch is not distinguishable from any other failure by a passive
// observer.
//
// # Limitations
//
// This protects the credentials in a request body. It is NOT a substitute for
// TLS: the session token the server returns afterwards, and every other byte
// of the response, still travel in the clear. Deploy behind HTTPS where
// possible; this is a mitigation for the cases where that is not an option.
package credcrypt

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"
)

// Algorithm is the identifier reported to the client. It names the key-wrapping
// algorithm; the payload cipher (AES-256-GCM) is implied by the envelope shape.
const Algorithm = "RSA-OAEP-256"

// DefaultChallengeTTL is how long a handshake challenge stays usable. It only
// has to cover the round trip between the handshake and the request it belongs
// to, so it is deliberately short.
const DefaultChallengeTTL = 2 * time.Minute

const (
	// rsaKeyBits is the modulus size of the ephemeral key pair.
	rsaKeyBits = 2048
	// aesKeyBytes is the size of the generated payload key: AES-256.
	aesKeyBytes = 32
	// challengeBytes is the entropy of a challenge token. It only needs to be
	// unguessable within the TTL, but guessing it is not even the main defence
	// — a challenge is consumed on first use.
	challengeBytes = 24
	// maxChallenges caps the challenge table. Reaching it means either a flood
	// of handshakes or a client that never finishes logging in; both are
	// handled by pruning expired entries and then refusing new ones, so the
	// table cannot be used to exhaust memory.
	maxChallenges = 1 << 14
)

// Errors returned by Open. Callers map them onto responses; the distinction
// that matters is whether retrying with a fresh handshake could succeed.
var (
	// ErrStaleKey means the envelope was encrypted to a public key this server
	// no longer holds — typically a cached key from before a restart. A fresh
	// handshake fixes it.
	ErrStaleKey = errors.New("credcrypt: envelope was encrypted to a stale public key")

	// ErrChallenge means the challenge was unknown, already used, expired, or
	// issued for a different scope. A fresh handshake is required.
	ErrChallenge = errors.New("credcrypt: challenge is missing, expired or already used")

	// ErrMalformed means the envelope is not well formed, its authentication
	// tag does not verify, or its payload cannot be parsed. No amount of
	// retrying helps.
	ErrMalformed = errors.New("credcrypt: malformed envelope")

	// ErrExhausted is returned when the challenge table is full of live
	// challenges and no new one can be issued.
	ErrExhausted = errors.New("credcrypt: too many pending handshakes")
)

// Envelope is the encrypted replacement for a request body, as it appears on
// the wire:
//
//	{"encrypted": {"kid": "...", "k": "...", "iv": "...", "d": "..."}}
//
// All four fields are standard base64. KeyID lets the server reject a stale key
// before doing any asymmetric work; Key and IV are the RSA-wrapped AES key and
// the GCM nonce; Data is the sealed payload.
type Envelope struct {
	KeyID string `json:"kid"`
	Key   string `json:"k"`
	IV    string `json:"iv"`
	Data  string `json:"d"`
}

// encryptedPayload is the plaintext of an envelope, before it is encrypted. The
// application body is kept as raw JSON so that the server can hand it to the
// existing handlers unmodified, without knowing its schema.
type encryptedPayload struct {
	Challenge string          `json:"challenge"`
	Payload   json.RawMessage `json:"payload"`
}

type challenge struct {
	scope   string
	expires time.Time
}

// Service holds the ephemeral key pair and the outstanding challenges.
type Service struct {
	priv   *rsa.PrivateKey
	pubDER string // standard base64 of the SPKI encoding
	keyID  string // short fingerprint of pubDER

	ttl time.Duration

	mu         sync.Mutex
	challenges map[string]challenge
}

// New generates a key pair and returns a ready Service.
//
// The key pair lives only in memory: it is created at startup and replaced on
// every restart. That keeps the private key off disk entirely, and the cost is
// only that a client holding a cached public key has to redo the handshake
// after a restart — which the KeyID check makes cheap and explicit.
func New(ttl time.Duration) (*Service, error) {
	if ttl <= 0 {
		ttl = DefaultChallengeTTL
	}

	priv, err := rsa.GenerateKey(rand.Reader, rsaKeyBits)
	if err != nil {
		return nil, fmt.Errorf("credcrypt: generate key: %w", err)
	}

	der, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
	if err != nil {
		return nil, fmt.Errorf("credcrypt: marshal public key: %w", err)
	}

	sum := sha256.Sum256(der)

	return &Service{
		priv:       priv,
		pubDER:     base64.StdEncoding.EncodeToString(der),
		keyID:      base64.RawURLEncoding.EncodeToString(sum[:9]),
		ttl:        ttl,
		challenges: make(map[string]challenge),
	}, nil
}

// TTL reports how long a freshly issued challenge remains valid.
func (s *Service) TTL() time.Duration { return s.ttl }

// KeyID is the short, non-secret identifier of the current public key. Clients
// echo it back in an envelope so the server can detect a stale key without
// attempting a decryption that is bound to fail.
func (s *Service) KeyID() string { return s.keyID }

// PublicKey returns the base64 SPKI encoding of the public key, as accepted by
// WebCrypto's importKey("spki", ...) and by node-forge's
// pki.publicKeyFromAsn1 after a DER decode.
func (s *Service) PublicKey() string { return s.pubDER }

// Challenge issues a single-use token for the given scope.
func (s *Service) Challenge(scope string) (string, error) {
	buf := make([]byte, challengeBytes)
	if _, err := rand.Read(buf); err != nil {
		return "", fmt.Errorf("credcrypt: generate challenge: %w", err)
	}
	token := base64.RawURLEncoding.EncodeToString(buf)

	now := time.Now()

	s.mu.Lock()
	defer s.mu.Unlock()

	if len(s.challenges) >= maxChallenges {
		s.sweepLocked(now)
	}
	if len(s.challenges) >= maxChallenges {
		return "", ErrExhausted
	}

	s.challenges[token] = challenge{scope: scope, expires: now.Add(s.ttl)}
	return token, nil
}

// Open validates an envelope and returns the application body it carries.
//
// The returned bytes are the value of the envelope's "payload" property, ready
// to be used as a request body. Order matters here: the challenge is only
// consumed once the payload is known to be authentic, so a forged envelope
// cannot burn someone else's challenge.
func (s *Service) Open(scope string, env Envelope) (json.RawMessage, error) {
	if env.KeyID != s.keyID {
		return nil, ErrStaleKey
	}

	wrapped, err := base64.StdEncoding.DecodeString(env.Key)
	if err != nil {
		return nil, ErrMalformed
	}
	iv, err := base64.StdEncoding.DecodeString(env.IV)
	if err != nil {
		return nil, ErrMalformed
	}
	sealed, err := base64.StdEncoding.DecodeString(env.Data)
	if err != nil {
		return nil, ErrMalformed
	}

	key, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, s.priv, wrapped, nil)
	if err != nil {
		// The envelope named the current key but was not wrapped with it:
		// either it was tampered with or the client and server disagree about
		// the key. Both are answered by redoing the handshake.
		return nil, ErrStaleKey
	}
	if len(key) != aesKeyBytes {
		return nil, ErrMalformed
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, ErrMalformed
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, ErrMalformed
	}
	if len(iv) != gcm.NonceSize() {
		return nil, ErrMalformed
	}

	plain, err := gcm.Open(nil, iv, sealed, nil)
	if err != nil {
		// Authentication failed: wrong key, truncated, or modified ciphertext.
		return nil, ErrMalformed
	}

	var inner encryptedPayload
	if err := json.Unmarshal(plain, &inner); err != nil {
		return nil, ErrMalformed
	}
	if len(inner.Payload) == 0 || !json.Valid(inner.Payload) {
		return nil, ErrMalformed
	}

	if !s.consume(inner.Challenge, scope) {
		return nil, ErrChallenge
	}

	return inner.Payload, nil
}

// consume redeems a challenge: it succeeds only if the token is outstanding,
// unexpired, and was issued for this scope, and it never succeeds twice.
func (s *Service) consume(token, scope string) bool {
	if token == "" {
		return false
	}

	now := time.Now()

	s.mu.Lock()
	defer s.mu.Unlock()

	ch, ok := s.challenges[token]
	if !ok {
		return false
	}
	// Delete unconditionally, so a wrong scope or an expired token still burns
	// it and cannot be probed repeatedly.
	delete(s.challenges, token)

	if ch.scope != scope || now.After(ch.expires) {
		return false
	}

	return true
}

// sweepLocked drops expired challenges. It must be called with s.mu held.
func (s *Service) sweepLocked(now time.Time) {
	for token, ch := range s.challenges {
		if now.After(ch.expires) {
			delete(s.challenges, token)
		}
	}
}

// Pending reports how many challenges are outstanding. It exists for tests and
// diagnostics.
func (s *Service) Pending() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.challenges)
}

File: frontend/src/utils/credcrypt.ts
import forge from "node-forge";
import { baseURL } from "./constants";

/**
 * Credential encryption for File Browser deployments served over plain HTTP.
 *
 * ## Why this exists
 *
 * A login POST carries a password. Over `http://` that password is readable by
 * anyone able to watch the connection — a shared Wi-Fi segment, a mirroring
 * switch port, a logging reverse proxy. TLS is the real answer, but a file
 * server on a home or office LAN often cannot have it, and the password people
 * type there is usually the one they reuse elsewhere.
 *
 * ## How it works
 *
 * 1. The client asks `/api/crypto/handshake` for the server's ephemeral RSA
 *    public key and a single-use challenge for the operation it is about to
 *    perform.
 * 2. It generates a random AES-256-GCM key, seals the request payload with it,
 *    and wraps that key with RSA-OAEP (SHA-256) under the server's key.
 * 3. It POSTs `{"encrypted": {kid, k, iv, d}}` in place of the plaintext body.
 *
 * Only the server can unwrap the payload, so the password is never on the wire.
 * The challenge inside the sealed payload is what stops the ciphertext from
 * being replayed verbatim by the same passive attacker who captured it: it is
 * accepted exactly once, only for the scope it was issued for, and only until it
 * expires.
 *
 * ## Why node-forge and not WebCrypto
 *
 * `crypto.subtle` is only exposed in a *secure context*. `http://localhost`
 * qualifies, but `http://192.168.1.10:8080` — the deployment this whole module
 * exists for — does not. WebCrypto is therefore unavailable exactly where it is
 * needed, so the primitives come from a bundled pure-JS library instead. This
 * costs some bundle size; it is the price of working on the deployments that
 * need it.
 *
 * ## What this does not protect
 *
 * The session token the server returns after login, and every other byte of
 * every response, still travel in the clear. Deploy behind HTTPS where that is
 * possible.
 */

interface Handshake {
  algorithm: string;
  keyID: string;
  key: string;
  challenge: string;
  expiresIn: number;
}

/** The encrypted replacement for a request body, as it goes on the wire. */
export interface Envelope {
  kid: string;
  k: string;
  iv: string;
  d: string;
}

export interface EncryptedBody {
  encrypted: Envelope;
}

const AES_KEY_BYTES = 32;
const GCM_NONCE_BYTES = 12;
const GCM_TAG_BITS = 128;

/**
 * Raised when the handshake itself fails, i.e. before any credential has been
 * sent. The API layer turns this into a StatusError so the UI reports it
 * normally.
 */
export class HandshakeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "HandshakeError";
  }
}

/**
 * Fetches a public key and a challenge for one operation.
 *
 * A fresh challenge is required for every request, so this is deliberately not
 * cached; the key could be, but the round trip is what makes replay detection
 * possible and it costs one small request next to the login it protects.
 */
async function handshake(scope: string): Promise<Handshake> {
  const res = await fetch(`${baseURL}/api/crypto/handshake`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ scope }),
  });

  if (!res.ok) {
    throw new HandshakeError(
      `credential handshake failed: ${res.status} ${res.statusText}`
    );
  }

  const data = (await res.json()) as Handshake;
  if (!data.key || !data.challenge) {
    throw new HandshakeError("credential handshake returned no key");
  }

  return data;
}

/**
 * Rebuilds a forge public key from the base64 SubjectPublicKeyInfo the server
 * hands out — the same encoding `x509.MarshalPKIXPublicKey` produces.
 */
function parsePublicKey(spkiBase64: string): forge.pki.rsa.PublicKey {
  const der = forge.util.decode64(spkiBase64);
  const asn1 = forge.asn1.fromDer(der);
  const key = forge.pki.publicKeyFromAsn1(asn1);

  if (typeof key === "string") {
    throw new HandshakeError("server sent an unsupported public key");
  }

  return key;
}

/**
 * Seals `payload` for the given scope and returns the envelope to send in its
 * place.
 */
export async function encryptBody(
  scope: string,
  payload: unknown
): Promise<EncryptedBody> {
  const { keyID, key, challenge } = await handshake(scope);
  const publicKey = parsePublicKey(key);

  const aesKey = forge.random.getBytesSync(AES_KEY_BYTES);
  const iv = forge.random.getBytesSync(GCM_NONCE_BYTES);

  const cipher = forge.cipher.createCipher("AES-GCM", aesKey);
  cipher.start({ iv, tagLength: GCM_TAG_BITS });
  cipher.update(
    forge.util.createBuffer(
      forge.util.encodeUtf8(JSON.stringify({ challenge, payload }))
    )
  );

  if (!cipher.finish()) {
    throw new HandshakeError("failed to encrypt the request payload");
  }

  // Go's cipher.AEAD expects the tag appended to the ciphertext.
  const sealed = cipher.output.getBytes() + cipher.mode.tag.getBytes();

  const wrappedKey = publicKey.encrypt(aesKey, "RSA-OAEP", {
    md: forge.md.sha256.create(),
    mgf1: { md: forge.md.sha256.create() },
  });

  return {
    encrypted: {
      kid: keyID,
      k: forge.util.encode64(wrappedKey),
      iv: forge.util.encode64(iv),
      d: forge.util.encode64(sealed),
    },
  };
}

/**
 * Encrypts a single secret for transport in a header, such as a share's
 * `X-SHARE-PASSWORD`. The wire form is the literal prefix `enc:` followed by the
 * envelope as JSON, which base64 cannot be confused with.
 */
export async function encryptHeaderValue(
  scope: string,
  value: string
): Promise<string> {
  const body = await encryptBody(scope, { value });
  return `enc:${JSON.stringify(body.encrypted)}`;
}

/**
 * True when the server rejected an envelope that needs to be rebuilt: either the
 * challenge was already used or has expired, or the public key changed under us
 * (the key pair is ephemeral, so a server restart invalidates cached keys).
 */
export function isStaleEnvelope(status: number | undefined): boolean {
  return status === 428;
}

/* ------------------------------------------------------------------ sessions
 *
 * A JWT is a bearer token: anyone who reads one off the wire can use it. So the
 * browser stops sending it. At login it generates a session key, sends it up
 * inside the already encrypted login envelope, and gets back an encrypted reply
 * carrying the token and a *sealed* copy of that session key. From then on every
 * request proves possession of the session key instead of presenting the token:
 *
 *   v2.<sealed session key>.<unix seconds>.<nonce>.<AES-GCM(session key, JWT)>
 *
 * Nothing in that header can be turned into a usable credential by someone who
 * captured it. The session key is inside the sealed blob, which only the server
 * can open; the nonce is part of the AEAD's additional data, so it cannot be
 * swapped; and the server spends each nonce exactly once, so the ciphertext
 * cannot be replayed.
 *
 * This is a defence against a passive eavesdropper. It does nothing against an
 * active attacker — see the package note above.
 * ------------------------------------------------------------------------- */

const JWT_STORAGE_KEY = "jwt";
const SESSION_KEY_STORAGE_KEY = "sessionKey";
const SEALED_KEY_STORAGE_KEY = "sealedKey";

const SESSION_KEY_BYTES = 32;
const CREDENTIAL_NONCE_BYTES = 12;
const CREDENTIAL_AAD_PREFIX = "filebrowser/credential/v1\n";
const RESPONSE_AAD = "filebrowser/session-response/v1";

/** What the client needs to authenticate a request. */
export interface CredentialMaterial {
  jwt: string;
  sessionKey: string;
  sealedKey: string;
}

/** A freshly generated session key, base64, as the login payload carries it. */
export function generateSessionKey(): string {
  return forge.util.encode64(forge.random.getBytesSync(SESSION_KEY_BYTES));
}

/** The credential material currently in storage, empty strings when absent. */
export function credentialMaterial(): CredentialMaterial {
  return {
    jwt: localStorage.getItem(JWT_STORAGE_KEY) ?? "",
    sessionKey: localStorage.getItem(SESSION_KEY_STORAGE_KEY) ?? "",
    sealedKey: localStorage.getItem(SEALED_KEY_STORAGE_KEY) ?? "",
  };
}

export function storeCredentialMaterial(material: CredentialMaterial) {
  localStorage.setItem(JWT_STORAGE_KEY, material.jwt);
  localStorage.setItem(SESSION_KEY_STORAGE_KEY, material.sessionKey);
  localStorage.setItem(SEALED_KEY_STORAGE_KEY, material.sealedKey);
}

export function clearCredentialMaterial() {
  localStorage.setItem(JWT_STORAGE_KEY, "");
  localStorage.setItem(SESSION_KEY_STORAGE_KEY, "");
  localStorage.setItem(SEALED_KEY_STORAGE_KEY, "");
}

/** base64url without padding, matching Go's base64.RawURLEncoding. */
function toBase64Url(value: string): string {
  return forge.util.encode64(value).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

/** Decodes a base64 session key into the raw key bytes forge works with. */
function sessionCipher(sessionKeyBase64: string): string {
  return forge.util.decode64(sessionKeyBase64);
}

/**
 * Builds the v2 credential for one request, or "" when there is nothing to
 * authenticate with.
 *
 * The method is bound into the tag, so a credential minted for a read cannot be
 * pointed at a write. The exact request target deliberately is not: reconciling
 * the browser's URL serialiser with Go's would make any filename containing an
 * unusual character fail to authenticate, and the nonce already rules out reuse.
 */
export function buildCredential(method: string, material: CredentialMaterial): string {
  if (!material.jwt || !material.sessionKey || !material.sealedKey) {
    return "";
  }

  const key = sessionCipher(material.sessionKey);
  const nonce = forge.random.getBytesSync(CREDENTIAL_NONCE_BYTES);
  const nonceText = toBase64Url(nonce);
  const issuedAt = Math.floor(Date.now() / 1000);

  const cipher = forge.cipher.createCipher("AES-GCM", key);
  cipher.start({
    iv: nonce,
    tagLength: 128,
    additionalData: `${CREDENTIAL_AAD_PREFIX}${method}\n${issuedAt}\n${nonceText}`,
  });
  cipher.update(forge.util.createBuffer(material.jwt));

  if (!cipher.finish()) {
    throw new HandshakeError("failed to build the request credential");
  }

  const ciphertext = cipher.output.getBytes() + cipher.mode.tag.getBytes();

  return [
    "v2",
    material.sealedKey,
    issuedAt,
    nonceText,
    forge.util.encode64(ciphertext),
  ].join(".");
}

/** A token reply, as the server sends it: an envelope over the session key. */
interface EncryptedTokenReply {
  encrypted: { iv: string; d: string };
}

function isEncryptedReply(value: unknown): value is EncryptedTokenReply {
  const candidate = value as EncryptedTokenReply | null;
  return Boolean(candidate?.encrypted?.iv && candidate.encrypted.d);
}

/**
 * Reads a token reply, decrypting it when the server encrypted it.
 *
 * A client that sent no session key still gets a plaintext token, which is what
 * keeps the CLI and third-party clients working.
 */
export function readTokenReply(
  contentType: string | null,
  body: string,
  sessionKey: string
): { token: string; sealedKey: string } {
  if (!contentType?.includes("application/json")) {
    return { token: body, sealedKey: "" };
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(body);
  } catch {
    throw new HandshakeError("the server sent an unreadable token reply");
  }

  if (!isEncryptedReply(parsed)) {
    return { token: body, sealedKey: "" };
  }

  const key = sessionCipher(sessionKey);
  const iv = forge.util.decode64(parsed.encrypted.iv);
  const sealed = forge.util.decode64(parsed.encrypted.d);

  // Go's cipher.AEAD appends the tag to the ciphertext.
  const tagLength = 16;
  const tag = sealed.slice(sealed.length - tagLength);
  const ciphertext = sealed.slice(0, sealed.length - tagLength);

  const decipher = forge.cipher.createDecipher("AES-GCM", key);
  decipher.start({
    iv,
    tagLength: 128,
    tag: forge.util.createBuffer(tag),
    additionalData: RESPONSE_AAD,
  });
  decipher.update(forge.util.createBuffer(ciphertext));

  if (!decipher.finish()) {
    throw new HandshakeError("the server's token reply did not authenticate");
  }

  const payload = JSON.parse(decipher.output.toString()) as {
    token: string;
    sealedKey: string;
  };

  return { token: payload.token, sealedKey: payload.sealedKey };
}
